From c86ee9f125bde2c81956934f16d5d179aa88a364 Mon Sep 17 00:00:00 2001
From: Ciprian Marian Costea <ciprianmarian.costea@nxp.com>
Date: Tue, 17 Aug 2021 14:31:08 +0300
Subject: [PATCH 07/19] s32gen1: flexcan: Fix unexpected unlock

commit c86ee9f125bde2c81956934f16d5d179aa88a364 from
https://source.codeaurora.org/external/autobsps32/linux

Before this change, in 'flexcan_mailbox_read' if the message was
dropped in the first check, execution will go to 'mark_as_read' label,
where it will unlock a never previously acquired lock.

Upstream-Status: Pending

Issue: ALB-7527
Signed-off-by: Ciprian Marian Costea <ciprianmarian.costea@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/can/flexcan.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/can/flexcan.c b/drivers/net/can/flexcan.c
index 758914ba3bc4..5b198f1b4228 100644
--- a/drivers/net/can/flexcan.c
+++ b/drivers/net/can/flexcan.c
@@ -1180,6 +1180,7 @@ static struct sk_buff *flexcan_mailbox_read(struct can_rx_offload *offload,
 
 	if (unlikely(drop)) {
 		skb = ERR_PTR(-ENOBUFS);
+		spin_lock_irqsave(&priv->timer_access, flags);
 		goto mark_as_read;
 	}
 
-- 
2.17.1

